docs/abi.rst

```diff
@@ -0,0 +1,560 @@
+            ),
+            InnerTxnBuilder.Submit(),
+        )
+
+Accessing Parameters of Referenced Values
+''''''''''''''''''''''''''''''''''''''''''
+
+Reference types allow the program to access more information about them. Each reference type has a :code:`params()` method which can be used to access that object's parameters. These methods are listed below:
+
+* :any:`abi.Account.params()` returns an :any:`AccountParamObject`
+* :any:`abi.Asset.params()` returns an :any:`AssetParamObject`
+* :any:`abi.Application.params()` returns an :any:`AppParamObject`
+
+These method are provided for convenience. They expose the same properties accessible from the :any:`AccountParam`, :any:`AssetParam`, and :any:`AppParam` classes.
+
+A brief example is below:
+
+.. code-block:: python
+
+    @Subroutine(TealType.none)
+    def referenced_params_example(
+        account: abi.Account, asset: abi.Asset, app: abi.Application
+    ) -> Expr:
+        return Seq(
+            account.params().auth_address().outputReducer(
+                lambda value, has_value: Assert(And(has_value, value == Global.zero_address()))
+            ),
+            asset.params().total().outputReducer(
+                lambda value, has_value: Assert(And(has_value, value == Int(1)))
+            ),
+            app.params().creator_address().outputReducer(
+                lambda value, has_value: Assert(And(has_value, value == Txn.sender()))
+            )
+        )
+
+.. note::
+    All returned parameters are instances of :any:`MaybeValue`.
+
+Accessing Asset Holdings
+''''''''''''''''''''''''
+
+Similar to the parameters above, asset holding properties can be accessed using one of the following methods:
+
+* :any:`abi.Account.asset_holding()`: given an asset, returns an :any:`AssetHoldingObject`
+* :any:`abi.Asset.holding()`: given an account, returns an :any:`AssetHoldingObject`
+
+These method are provided for convenience. They expose the same properties accessible from the :any:`AssetHolding` class.
+
+A brief example is below:
+
+.. code-block:: python
+
+    @Subroutine(TealType.none)
+    def ensure_asset_balance_is_nonzero(account: abi.Account, asset: abi.Asset) -> Expr:
+        return Seq(
+            account.asset_holding(asset)
+            .balance()
+            .outputReducer(lambda value, has_value: Assert(And(has_value, value > Int(0)))),
+            # this check is equivalent
+            asset.holding(account)
+            .balance()
+            .outputReducer(lambda value, has_value: Assert(And(has_value, value > Int(0)))),
+        )
+
+Limitations
+""""""""""""""""""""""""""""""""""""""""""
+
+TODO: explain limitations, such as can't be created directly, or used as method return value
+
+Transaction Types
+^^^^^^^^^^^^^^^^^^^^^^^^
+
+Some application calls require that they are invoked as part of a larger transaction group containing specific additional transactions. In order to express these types of calls, the ABI has transaction types.
```

Comment:
I was thinking about this in the Reference Types section as well, but would you consider capitalizing "Reference Types" and "Transaction Types"? I'm not sure myself, but it feels like lower case is a little harder to parse - it is that the ABI has transaction _types_, or that the ABI has _transaction_ types, or that the ABI has _transaction types_? Since these are both such common words in our day to day, it might be worth capitalizing.